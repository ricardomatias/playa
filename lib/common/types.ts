export enum BinaryEvent {
	Rest = 0,
	Hit = 1
}

export type Octaves = [number, number]

export enum RhythmType {
	Free = 'Free',
	Turn = 'Turn'
}

export type DistributionFunction = <T>(distribution: T | T[], precision?: number | undefined) => string[];
