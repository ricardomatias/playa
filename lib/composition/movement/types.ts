import { Interval, NoteSymbol, ScaleName, TurnMoves } from '../../constants';
import { GreekModeIntervals, ModePosition } from '../../constants/modes';

export const EUCLIDEAN = 'euclid';
export const TURN = 'turn';

export enum MovementRhythm {
	Euclidean = 'Euclidean',
	Turn = 'Turn'
}

export interface TimelineEventKey { root: NoteSymbol, scale: GreekModeIntervals }

export interface TimelineEvent {
	time: number;
	dur: number;
	key: TimelineEventKey;
}

export enum ModulationEventType {
	Mode = 'mode',
	Key = 'key'
}

export interface ModulationEvent {
	key: NoteSymbol;
	root: NoteSymbol;
	type: ModulationEventType;
	scaleName: ScaleName;
	position: ModePosition;
}

export type Movement = {
	timeline: TimelineEvent[];
	events: ModulationEvent[];
}

export interface Turn {
	type: TurnMoves,
	interval: typeof TurnMoves.Free | Interval | number | string
}

/**
 * @typedef {Object} Turn
 * @property {String} type '▼'
 * @property {String | Number} interval '⟷'
 * @example
 * { type: '▼', interval: '⟷' }
*/
export interface TurnEvent extends ModulationEvent {
	turn: Turn
}

export type TurnMovement = {
	timeline: TimelineEvent[];
	events: TurnEvent[];
}
