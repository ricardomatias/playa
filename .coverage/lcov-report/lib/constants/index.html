<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for lib/constants</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../index.html">All files</a> lib/constants
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">87.5% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>56/64</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">66.67% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>4/6</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">33.33% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>2/6</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">88.89% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>56/63</span>
      </div>
    </div>
    <p class="quiet">
      Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
    </p>
  </div>
  <div class='status-line high'></div>
<div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file high" data-value="chords.js"><a href="chords.js.html">chords.js</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="36" class="abs high">36/36</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="36" class="abs high">36/36</td>
	</tr>

<tr>
	<td class="file empty" data-value="index.js"><a href="index.js.html">index.js</a></td>
	<td data-value="0" class="pic empty"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	</tr>

<tr>
	<td class="file high" data-value="intervals.js"><a href="intervals.js.html">intervals.js</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	</tr>

<tr>
	<td class="file high" data-value="note.js"><a href="note.js.html">note.js</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="4" class="abs high">4/4</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="4" class="abs high">4/4</td>
	</tr>

<tr>
	<td class="file high" data-value="rhythmsDistributions.js"><a href="rhythmsDistributions.js.html">rhythmsDistributions.js</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	</tr>

<tr>
	<td class="file high" data-value="scales.js"><a href="scales.js.html">scales.js</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	</tr>

<tr>
	<td class="file medium" data-value="ticks.js"><a href="ticks.js.html">ticks.js</a></td>
	<td data-value="60" class="pic medium"><div class="chart"><div class="cover-fill" style="width: 60%;"></div><div class="cover-empty" style="width:40%;"></div></div></td>
	<td data-value="60" class="pct medium">60%</td>
	<td data-value="20" class="abs medium">12/20</td>
	<td data-value="66.67" class="pct medium">66.67%</td>
	<td data-value="6" class="abs medium">4/6</td>
	<td data-value="33.33" class="pct low">33.33%</td>
	<td data-value="6" class="abs low">2/6</td>
	<td data-value="63.16" class="pct medium">63.16%</td>
	<td data-value="19" class="abs medium">12/19</td>
	</tr>

</tbody>
</table>
</div><div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Wed Feb 20 2019 14:43:58 GMT+0100 (Central European Standard Time)
</div>
</div>
<script src="../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../sorter.js"></script>
<script src="../../block-navigation.js"></script>
</body>
</html>
