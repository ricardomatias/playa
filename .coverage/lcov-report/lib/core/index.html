<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for lib/core</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../index.html">All files</a> lib/core
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">88.17% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>343/389</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">81.98% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>141/172</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">83.82% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>57/68</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">87.83% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>332/378</span>
      </div>
    </div>
    <p class="quiet">
      Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
    </p>
  </div>
  <div class='status-line high'></div>
<div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file high" data-value="Chord.js"><a href="Chord.js.html">Chord.js</a></td>
	<td data-value="96.9" class="pic high"><div class="chart"><div class="cover-fill" style="width: 96%;"></div><div class="cover-empty" style="width:4%;"></div></div></td>
	<td data-value="96.9" class="pct high">96.9%</td>
	<td data-value="129" class="abs high">125/129</td>
	<td data-value="90.91" class="pct high">90.91%</td>
	<td data-value="44" class="abs high">40/44</td>
	<td data-value="94.44" class="pct high">94.44%</td>
	<td data-value="18" class="abs high">17/18</td>
	<td data-value="96.77" class="pct high">96.77%</td>
	<td data-value="124" class="abs high">120/124</td>
	</tr>

<tr>
	<td class="file high" data-value="HarmonyBase.js"><a href="HarmonyBase.js.html">HarmonyBase.js</a></td>
	<td data-value="84.62" class="pic high"><div class="chart"><div class="cover-fill" style="width: 84%;"></div><div class="cover-empty" style="width:16%;"></div></div></td>
	<td data-value="84.62" class="pct high">84.62%</td>
	<td data-value="13" class="abs high">11/13</td>
	<td data-value="66.67" class="pct medium">66.67%</td>
	<td data-value="6" class="abs medium">4/6</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="4" class="abs high">4/4</td>
	<td data-value="84.62" class="pct high">84.62%</td>
	<td data-value="13" class="abs high">11/13</td>
	</tr>

<tr>
	<td class="file high" data-value="Key.js"><a href="Key.js.html">Key.js</a></td>
	<td data-value="94.67" class="pic high"><div class="chart"><div class="cover-fill" style="width: 94%;"></div><div class="cover-empty" style="width:6%;"></div></div></td>
	<td data-value="94.67" class="pct high">94.67%</td>
	<td data-value="75" class="abs high">71/75</td>
	<td data-value="83.33" class="pct high">83.33%</td>
	<td data-value="36" class="abs high">30/36</td>
	<td data-value="92.31" class="pct high">92.31%</td>
	<td data-value="13" class="abs high">12/13</td>
	<td data-value="94.67" class="pct high">94.67%</td>
	<td data-value="75" class="abs high">71/75</td>
	</tr>

<tr>
	<td class="file low" data-value="Mode.js"><a href="Mode.js.html">Mode.js</a></td>
	<td data-value="0" class="pic low"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="24" class="abs low">0/24</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="9" class="abs low">0/9</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="4" class="abs low">0/4</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="24" class="abs low">0/24</td>
	</tr>

<tr>
	<td class="file high" data-value="Note.js"><a href="Note.js.html">Note.js</a></td>
	<td data-value="89.74" class="pic high"><div class="chart"><div class="cover-fill" style="width: 89%;"></div><div class="cover-empty" style="width:11%;"></div></div></td>
	<td data-value="89.74" class="pct high">89.74%</td>
	<td data-value="78" class="abs high">70/78</td>
	<td data-value="90.48" class="pct high">90.48%</td>
	<td data-value="42" class="abs high">38/42</td>
	<td data-value="84.21" class="pct high">84.21%</td>
	<td data-value="19" class="abs high">16/19</td>
	<td data-value="89.61" class="pct high">89.61%</td>
	<td data-value="77" class="abs high">69/77</td>
	</tr>

<tr>
	<td class="file high" data-value="Scale.js"><a href="Scale.js.html">Scale.js</a></td>
	<td data-value="93.94" class="pic high"><div class="chart"><div class="cover-fill" style="width: 93%;"></div><div class="cover-empty" style="width:7%;"></div></div></td>
	<td data-value="93.94" class="pct high">93.94%</td>
	<td data-value="66" class="abs high">62/66</td>
	<td data-value="82.86" class="pct high">82.86%</td>
	<td data-value="35" class="abs high">29/35</td>
	<td data-value="80" class="pct high">80%</td>
	<td data-value="10" class="abs high">8/10</td>
	<td data-value="93.44" class="pct high">93.44%</td>
	<td data-value="61" class="abs high">57/61</td>
	</tr>

<tr>
	<td class="file empty" data-value="index.js"><a href="index.js.html">index.js</a></td>
	<td data-value="0" class="pic empty"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	</tr>

<tr>
	<td class="file high" data-value="types.js"><a href="types.js.html">types.js</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="4" class="abs high">4/4</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="4" class="abs high">4/4</td>
	</tr>

</tbody>
</table>
</div><div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Wed Feb 20 2019 15:09:45 GMT+0100 (Central European Standard Time)
</div>
</div>
<script src="../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../sorter.js"></script>
<script src="../../block-navigation.js"></script>
</body>
</html>
