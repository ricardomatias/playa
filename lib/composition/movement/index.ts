export { default as createMovement } from './default';
export { default as createTurnMovement } from './turn';
export * from './types';
