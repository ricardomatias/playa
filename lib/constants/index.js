/**
 * Constants
 * @namespace {module} Constants
 */
export * from './chords';
export * from './scales';
export * from './note';
export * from './turns';
export * from './time';
export { default as MIDI_NOTES } from './midi';
export { default as TICKS } from './ticks';
export { default as INTERVALS } from './intervals';
