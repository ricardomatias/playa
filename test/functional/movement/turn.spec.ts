import { TurnMoves } from '../../../lib/constants';
import { Key } from '../../../lib/core';
import { createTurnMovement } from '../../../lib/functional';
import { MovementRhythm } from '../../../lib/functional/movement/types';
import Random from '../../../lib/tools/random';


describe('A Movement test suite', () => {
	describe('Turn', () => {
		it('should generate a simple movement', () => {
			// given
			Random.setSeed('test');

			const aMaj = new Key('C', Key.Major);

			const turns = [
				{ type: TurnMoves.Start, interval: TurnMoves.Free },
				{ type: TurnMoves.ModeUp, interval: 5 },
				{ type: TurnMoves.Free, interval: TurnMoves.Free },
				{ type: TurnMoves.ModulateDown, interval: '5P' },
				{ type: TurnMoves.Start, interval: TurnMoves.Free },
				{ type: TurnMoves.ModeDown, interval: 5 },
				{ type: TurnMoves.ModulateUp, interval: '4P' },
				{ type: TurnMoves.ModeDown, interval: 5 },
			];

			// when
			const movement = createTurnMovement(aMaj, turns, '4:0:0');

			expect(movement.timeline).toHaveLength(8);
			expect(movement).toMatchSnapshot();
		});

		it('should generate non buggy', () => {
			// given
			Random.setSeed('massive-dragon-38');

			const cMaj = new Key('C', Key.Lydian);

			const turns = [
				{ type: TurnMoves.Start, interval: 5 },
				{ type: TurnMoves.ModeUp, interval: 4 },
				{ type: TurnMoves.Free, interval: 2 },
				{ type: TurnMoves.Keep, interval: 2 },
				{ type: TurnMoves.ModeUp, interval: 4 },
				{ type: TurnMoves.ModeDown, interval: 5 },
				{ type: TurnMoves.ModulateUp, interval: '2M' },
				{ type: TurnMoves.Free, interval: 2 },
			];

			// when
			const movement = createTurnMovement(cMaj, turns, '4:0:0');

			expect(movement.timeline).toHaveLength(8);
			expect(movement).toMatchSnapshot();
		});

		it('should generate a turn rhythm movement', () => {
			// given
			Random.setSeed('test');

			const aMin = new Key('A', Key.Minor);

			const turns = [
				{ type: TurnMoves.Start, interval: TurnMoves.Free },
				{ type: TurnMoves.ModeUp, interval: 5 },
				{ type: TurnMoves.Free, interval: TurnMoves.Free },
				{ type: TurnMoves.ModulateDown, interval: '5P' },
				{ type: TurnMoves.ModeUp, interval: 4 },
				{ type: TurnMoves.ModeDown, interval: 5 },
			];

			// when
			const movement = createTurnMovement(aMin, turns, '4:0:0', { rhythmType: MovementRhythm.Turn });

			expect(movement.timeline).toHaveLength(6);
			expect(movement).toMatchSnapshot();
		});
	});
});
