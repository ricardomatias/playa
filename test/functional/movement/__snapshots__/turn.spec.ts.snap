// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A Movement test suite Turn should generate a simple movement 1`] = `
Object {
  "events": Array [
    Object {
      "key": "C",
      "position": "II",
      "root": "D",
      "scaleName": "Dorian",
      "turn": Object {
        "interval": "Free",
        "type": "Start",
      },
      "type": "mode",
    },
    Object {
      "key": "C",
      "position": "VI",
      "root": "A",
      "scaleName": "Aeolian",
      "turn": Object {
        "interval": 5,
        "type": "ModeUp",
      },
      "type": "mode",
    },
    Object {
      "key": "C",
      "position": "III",
      "root": "E",
      "scaleName": "Phrygian",
      "turn": Object {
        "interval": "Free",
        "type": "Free",
      },
      "type": "mode",
    },
    Object {
      "key": "A",
      "position": "I",
      "root": "A",
      "scaleName": "Phrygian",
      "turn": Object {
        "interval": "5P",
        "type": "ModulateDown",
      },
      "type": "key",
    },
    Object {
      "key": "C",
      "position": "VII",
      "root": "B",
      "scaleName": "Locrian",
      "turn": Object {
        "interval": "Free",
        "type": "Start",
      },
      "type": "mode",
    },
    Object {
      "key": "C",
      "position": "III",
      "root": "E",
      "scaleName": "Phrygian",
      "turn": Object {
        "interval": 5,
        "type": "ModeDown",
      },
      "type": "mode",
    },
    Object {
      "key": "A",
      "position": "I",
      "root": "A",
      "scaleName": "Phrygian",
      "turn": Object {
        "interval": "4P",
        "type": "ModulateUp",
      },
      "type": "key",
    },
    Object {
      "key": "A",
      "position": "IV",
      "root": "D",
      "scaleName": "Aeolian",
      "turn": Object {
        "interval": 5,
        "type": "ModeDown",
      },
      "type": "mode",
    },
  ],
  "timeline": Array [
    Object {
      "dur": 960,
      "key": Object {
        "root": "D",
        "scale": "1P 2M 3m 4P 5P 6M 7m",
      },
      "time": 0,
    },
    Object {
      "dur": 960,
      "key": Object {
        "root": "A",
        "scale": "1P 2M 3m 4P 5P 6m 7m",
      },
      "time": 960,
    },
    Object {
      "dur": 960,
      "key": Object {
        "root": "E",
        "scale": "1P 2m 3m 4P 5P 6m 7m",
      },
      "time": 1920,
    },
    Object {
      "dur": 960,
      "key": Object {
        "root": "A",
        "scale": "1P 2m 3m 4P 5P 6m 7m",
      },
      "time": 2880,
    },
    Object {
      "dur": 960,
      "key": Object {
        "root": "B",
        "scale": "1P 2m 3m 4P 5d 6m 7m",
      },
      "time": 3840,
    },
    Object {
      "dur": 960,
      "key": Object {
        "root": "E",
        "scale": "1P 2m 3m 4P 5P 6m 7m",
      },
      "time": 4800,
    },
    Object {
      "dur": 960,
      "key": Object {
        "root": "A",
        "scale": "1P 2m 3m 4P 5P 6m 7m",
      },
      "time": 5760,
    },
    Object {
      "dur": 960,
      "key": Object {
        "root": "D",
        "scale": "1P 2M 3m 4P 5P 6m 7m",
      },
      "time": 6720,
    },
  ],
}
`;

exports[`A Movement test suite Turn should generate a turn rhythm movement 1`] = `
Object {
  "events": Array [
    Object {
      "key": "A",
      "position": "IV",
      "root": "D",
      "scaleName": "Dorian",
      "turn": Object {
        "interval": "Free",
        "type": "Start",
      },
      "type": "mode",
    },
    Object {
      "key": "A",
      "position": "I",
      "root": "A",
      "scaleName": "Aeolian",
      "turn": Object {
        "interval": 5,
        "type": "ModeUp",
      },
      "type": "mode",
    },
    Object {
      "key": "A",
      "position": "IV",
      "root": "D",
      "scaleName": "Dorian",
      "turn": Object {
        "interval": "Free",
        "type": "Free",
      },
      "type": "mode",
    },
    Object {
      "key": "G",
      "position": "I",
      "root": "G",
      "scaleName": "Dorian",
      "turn": Object {
        "interval": "5P",
        "type": "ModulateDown",
      },
      "type": "key",
    },
    Object {
      "key": "G",
      "position": "IV",
      "root": "C",
      "scaleName": "Mixolydian",
      "turn": Object {
        "interval": 4,
        "type": "ModeUp",
      },
      "type": "mode",
    },
    Object {
      "key": "G",
      "position": "VII",
      "root": "F",
      "scaleName": "Ionian",
      "turn": Object {
        "interval": 5,
        "type": "ModeDown",
      },
      "type": "mode",
    },
  ],
  "timeline": Array [
    Object {
      "dur": 1280,
      "key": Object {
        "root": "D",
        "scale": "1P 2M 3m 4P 5P 6M 7m",
      },
      "time": 0,
    },
    Object {
      "dur": 1440,
      "key": Object {
        "root": "A",
        "scale": "1P 2M 3m 4P 5P 6m 7m",
      },
      "time": 1280,
    },
    Object {
      "dur": 1440,
      "key": Object {
        "root": "D",
        "scale": "1P 2M 3m 4P 5P 6M 7m",
      },
      "time": 2720,
    },
    Object {
      "dur": 1280,
      "key": Object {
        "root": "G",
        "scale": "1P 2M 3m 4P 5P 6M 7m",
      },
      "time": 4160,
    },
    Object {
      "dur": 1280,
      "key": Object {
        "root": "C",
        "scale": "1P 2M 3M 4P 5P 6M 7m",
      },
      "time": 5440,
    },
    Object {
      "dur": 960,
      "key": Object {
        "root": "F",
        "scale": "1P 2M 3M 4P 5P 6M 7M",
      },
      "time": 6720,
    },
  ],
}
`;

exports[`A Movement test suite Turn should generate non buggy 1`] = `
Object {
  "events": Array [
    Object {
      "key": "C",
      "position": "V",
      "root": "G",
      "scaleName": "Ionian",
      "turn": Object {
        "interval": 5,
        "type": "Start",
      },
      "type": "mode",
    },
    Object {
      "key": "C",
      "position": "I",
      "root": "C",
      "scaleName": "Lydian",
      "turn": Object {
        "interval": 4,
        "type": "ModeUp",
      },
      "type": "mode",
    },
    Object {
      "key": "A",
      "position": "I",
      "root": "A",
      "scaleName": "Lydian",
      "turn": Object {
        "interval": "6M",
        "type": "ModulateUp",
      },
      "type": "key",
    },
    Object {
      "key": "A",
      "position": "I",
      "root": "A",
      "scaleName": "Lydian",
      "turn": Object {
        "interval": 2,
        "type": "Keep",
      },
      "type": "mode",
    },
    Object {
      "key": "A",
      "position": "IV",
      "root": "D#",
      "scaleName": "Locrian",
      "turn": Object {
        "interval": 4,
        "type": "ModeUp",
      },
      "type": "mode",
    },
    Object {
      "key": "A",
      "position": "VII",
      "root": "G#",
      "scaleName": "Phrygian",
      "turn": Object {
        "interval": 5,
        "type": "ModeDown",
      },
      "type": "mode",
    },
    Object {
      "key": "A#",
      "position": "I",
      "root": "A#",
      "scaleName": "Phrygian",
      "turn": Object {
        "interval": "2M",
        "type": "ModulateUp",
      },
      "type": "key",
    },
    Object {
      "key": "A#",
      "position": "I",
      "root": "A#",
      "scaleName": "Phrygian",
      "turn": Object {
        "interval": 2,
        "type": "Free",
      },
      "type": "mode",
    },
  ],
  "timeline": Array [
    Object {
      "dur": 960,
      "key": Object {
        "root": "G",
        "scale": "1P 2M 3M 4P 5P 6M 7M",
      },
      "time": 0,
    },
    Object {
      "dur": 960,
      "key": Object {
        "root": "C",
        "scale": "1P 2M 3M 4A 5P 6M 7M",
      },
      "time": 960,
    },
    Object {
      "dur": 960,
      "key": Object {
        "root": "A",
        "scale": "1P 2M 3M 4A 5P 6M 7M",
      },
      "time": 1920,
    },
    Object {
      "dur": 960,
      "key": Object {
        "root": "A",
        "scale": "1P 2M 3M 4A 5P 6M 7M",
      },
      "time": 2880,
    },
    Object {
      "dur": 960,
      "key": Object {
        "root": "D#",
        "scale": "1P 2m 3m 4P 5d 6m 7m",
      },
      "time": 3840,
    },
    Object {
      "dur": 960,
      "key": Object {
        "root": "G#",
        "scale": "1P 2m 3m 4P 5P 6m 7m",
      },
      "time": 4800,
    },
    Object {
      "dur": 960,
      "key": Object {
        "root": "A#",
        "scale": "1P 2m 3m 4P 5P 6m 7m",
      },
      "time": 5760,
    },
    Object {
      "dur": 960,
      "key": Object {
        "root": "A#",
        "scale": "1P 2m 3m 4P 5P 6m 7m",
      },
      "time": 6720,
    },
  ],
}
`;
